services/user_service/pubsub.py

```diff
@@ -0,0 +1,33 @@
+import os
+from google.cloud import pubsub_v1
+
+os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = os.getcwd() + "/google_cloud.json"
+
+
+class PubSubManager:
+    def __init__(self, sub_path, project_id):
+        self.__subscriber = pubsub_v1.SubscriberClient()
+        self.__publisher = pubsub_v1.PublisherClient()
+        self.__subscriber_path = sub_path
+        self.__project_id = project_id
+
+    def subscribe(self, subscription, callback):
+        path = self.__subscriber_path + subscription
+
+        print(f"Listening messages on {path}")
+
+        user_pull = self.__subscriber.subscribe(path, callback=callback)
+
+        with self.__subscriber:
+            try:
+                user_pull.result()
+            except:
+                user_pull.cancel()
+                user_pull.result()
```

Comment:
I think it should be in separate method, sth like "start_pulling()"
Do you know how it is working? In pulls infinitely until subscriber is valid?
Let's have methods for start and stop pulling

Reply:
this piece of code will never be duplicated or used  again. There is no need to keep a separate function for it. Regarding "how it works" I will answer later. Need to read some docs since I simply used an official sample.

Reply:
_this piece of code will never be duplicated or used again._
And what? Imagine function
```
func cook_breakfast(){
    make_coffee();
    make_tosts();
    go_to_work();
    return_home();
}
```
Even if last two statements won't be duplicated, they are not suitable for cooking
There **is** need to create separate function for pulling messages. As well as function for stopping pulling, which should be called from destructor

Reply:
Your example is correct but can not be applied to subscribe function. When you mess up diffent logic in one place like make_coffee and go_to_work in the func cool_breakfest, it's semantically wrong since does not respond for what function name declare to do.

Pooling naturally is the main part of subscription and logically responds to the function name. It does just one unit of work - subscribe.

I would leave it as it is.

Reply:
https://cloud.google.com/pubsub/docs/pull

Reply:
@ZhavoronkovN this is a python, we do not play with desctructors

Reply:
Subscribing and pulling are already semantically different things
Example you gave doesn't even use classes, but still you've created dedicated class for it

Reply:
Well python developers play with desctructors
That's why they created dedicated method for it

Reply:
I did refactor this, but it is disgusting for me (sorry).

Reply:
> I do not see here something important we should care about.

That's why there are other team members

Reply:
> 3\. No, they are not. When you subscribe to something it means you are waiting (pooling) for the messages

Implementation you are using have separate methods for subscribing and pulling. As well as any other implementation. Why do you think it is the same?

Reply:
@ZhavoronkovN i think that garbage collector will take of it, since we will not have any references to the object.

Reply:
We can rename the method to smth like `setup_subsriber` and this will handle this. Also, we can shut down the whole app when a thread fails.

Reply:
How?